### PR TITLE
Dependency API

### LINKED ISSUE #948: [RFC] Dependency API
**By: Schuyler Eldridge**

## History

| Date          | Version       |
|---------------|---------------|
| Nov 25th 2018 | Initial Draft |
| Nov 27th 2018 | Add LLVM PassManager Info |
| Dec 5th 2018 | Clarify examples, add Steve Burns prerequisite injection example |

## Motivation

The FIRRTL compiler middle end is the serial application of a sequence of `firrtl.Transform`. Each `Transform` is abstract in an `execute` method with signature:

```scala
abstract class Transform {
  def inputForm: CircuitForm
  def outputForm: CircuitForm

  /* Each Transform is abstract in an execute method */
  def execute(state: CircuitState): CircuitState
}

/* The CircuitState describes a snapshot of a Circuit and some metadata */
case class CircuitState(circuit: Circuit, form: CircuitForm, annotations: AnnotationSeq, renames: Option[RenameMap])
```

The Annotations/Options refactor (Issue: #764, Selected PRs: #919, #945, #920) generalizes the concept of a `Transform` as a `Phase`. A `Phase` is abstract in a transformation of an `AnnotationSeq` which allows this to generalize to components of the Chisel/FIRRTL Hardware Compiler Framework _not_ inside the FIRRTL middle end. Additionally, #945 introduces `Stage`, a subclass of `Phase`, that exposes a user interface. _Going forward, this CIP (Chisel Improvement Process) refers to `Phase` as if it were the superclass of `Transform` and `Pass` even though that is not currently the case._ A high-level overview of `Phase` and `Stage` are outlined in the code examples below:

```scala
abstract class Phase {
  /** Mathematical transformation of an [[AnnotationSeq]] */
  def transform(annotations: AnnotationSeq): AnnotationSeq
}

abstract class stage {
  /** A command line interface utility */
  val shell: Shell

  /** Abstract transformation that a subclass implements */
  def run(annotations: AnnotationSeq): AnnotationSeq

  /** Does common preprocessing and then calls the run method */
  final def transform(annotations: AnnotationSeq): AnnotationSeq = {
    val x = // preprocessing of input `annotations`
    run(x)
  }

  /** Method that allows for both command line arguments and annotations as inputs */
  final def execute(args: Array[String], annotations: AnnotationSeq): AnnotationSeq =
    transform(shell.parse(args, annotations))

  /* Command line interface API */
  final def main(args: Array[String]): Unit = execute(args, Seq.empty)
}
```

The discussion around `Stage` and `Phase` revived a long standing issue first brought up in #446:

> __How does the `Phase` scheduler derive the order in which to apply each `Phase`?__

Such an ordering is important for both:

1. An internal `Phase` that runs by default, e.g., `DeadCodeElimination`
2. An external, user-injected `Phase`, e.g., MIDAS' FAME1 transform

This CIP aims to develop two new APIs:

1. A __Dependency API__ for `Phase`
2. A __`Phase` Scheduler__ that derives a `Phase` order using (1)

At a high level, the __Dependency API__ needs to communicate two pieces of information for every `Phase`:

1. What are my prerequisites?
2. What do I update?

This information is mirrored by any build system or compiler. Consider [Make](https://en.wikipedia.org/wiki/Make_(software)). Make provides an API consisting of _rules_ and _recipes_. Each _rule_ defines a _target_ (something that this rule updates) and zero or more _prerequisites_. The _recipe_ is then called whenever any _prerequisite_ is newer than the _target_. Also consider [LLVM](http://llvm.org/docs/WritingAnLLVMPass.html#specifying-interactions-between-passes). Each pass in LLVM may optionally define a list of _prerequisite_ analyses that it needs and a list of analyses that it _invalidates_. Conceptually, the dependency API for both Make and LLVM are the same.

This CIP proposes that each `Phase` declares it's dependencies in the following way:

```scala
/** Defines the set of `Phase` that have been run */
case class PhaseStamps(phase: Set[Phase]) extends NoTargetAnnotation

abstract class Phase {
  /** A list of `Phase`s that must be run before this `Phase` */
  def prerequisites: Seq[Phase] = Seq.empty

  /** A function that is true if a particular `Phase` is invalidated by this `Phase` */
  def invalidate(phase: Phase): Boolean = true
}
```

First, this defines an `Annotation` that holds information about what `Phase`s have been run. Second, this defines the prerequisites that a specific `Phase` needs. Third, this defines a method `invalidate` that a manager can use to invalidate members of the set contained by the `PhaseStamps` annotation. By default, a `Phase` has no prerequisites and it invalidates everything.

Note: it may make sense to have the `PhaseStamps` be attached to something or a collection of something. E.g., one could consider a generalization of `FirrtlCircuitAnnotation` that, in addition to holding a `Circuit`, also holds the "state" of that circuit listing all `Phase`s that have been applied to it:
```scala
trait Stamped[T] { 
  val stamps: Set[Phase] 
  val data: Seq[T]
}

case class FirrtlCircuitAnnotation(data: Seq[Circuit], stamps: Set[Phase]) extends Stamped[Circuit]
```

This API defining `prerequisites` and `invalidates` is a generalization of the current `inputForm: CircuitForm` and `outputForm: CircuitForm`. The generalization being that it enables you to specify "partial" input and output forms and doesn't require the definition of a new `CircuitForm` and `LoweringCompiler`. This is critical if the `Phase` API is used outside of the FIRRTL middle end.

It then falls out naturally what a `PhaseManager` should do. A `PhaseManager` should, given an input `AnnotationSeq` and a `Set[Phase]` to run, determine a phase ordering (likely not optimal due to computational difficulty of doing so):

```scala
object PhaseManager {
  /** Run all necessary `Phase`s on an `AnnotationSeq` that executes all targets `Set[Phase]`
    * @throws PhaseManagerException if a cyclic Phase ordering was specified
    */
  def run(annotations: AnnotationSeq, targets: Set[Phase]): AnnotationSeq = {
    // TBD implementation
  }
}
```

# Motivating Examples

### Examples

See above for API.

### Comparison Examples

- How would `inputForm` and `outputForm` be migrated to this Dependency API?
- What is the equivalent operation for `firrtl.compilerUtils.getLoweringTransforms` and `firrtl.compilerUtils.mergeTransforms`?
- How would an injected dependency work, e.g., a user wants a specific `Phase` to run before another `Phase`? In effect, an injection of a new prerequisite for an existing `Phase` the user has not modified? (Is there a better way to do this?)

### Counter-Examples

I'm not sure what this _couldn't_ be used for. I'll need to think.

## Design

TBD.

### LLVM `PassManager`/`LegacyPassManager`

LLVM uses a `PassManager` to (by my initial skim) dynamically control the execution of passes and to error if something that looks like a cycle shows up. See [`PassManager.h`](https://github.com/llvm-mirror/llvm/blob/master/include/llvm/IR/PassManager.h#L443). 

The actual optimization appears to happen in `LegacyPassManager.cpp` in [`PMTopLevelManager::schedulePass`](https://github.com/llvm-mirror/llvm/blob/master/lib/IR/LegacyPassManager.cpp#L699).

## Counter-Examples

TBD

## Drawbacks

This is a big refactor...

TBD.

## Alternatives

TBD

## References

[1] [`PassManager.h`](https://github.com/llvm-mirror/llvm/blob/master/include/llvm/IR/PassManager.h#L443)

### PR DESCRIPTION
**Pull request by seldridge**
(Re-opening #1015 as Travis is not updating `$TRAVIS_BRANCH` after changing PR base branch...)

This implements a Dependency API for `TransformLike[A]` (hereafter referred to as **transforms**). This adds a `DependencyAPI` trait, an abstract class `DependencyManager`, and a concrete `PhaseManager`. The `DependencyAPI` defines three types of "dependencies" that, by implementing, allow a user to defer determination of a valid ordering to a `DependencyManager`. 

Dependencies of transform _Foo_ are then described using:

  - **prerequisites**: transforms that should run before Foo
  - **dependents**: transforms that should run after Foo (a means of prerequisite injection)
  - **invalidates**: transforms whose work/effects Foo destroys

Prerequisites and dependents are defined as a `Seq[Class[A]]` and internally converted to `LinkedHashSet[Class[A]]`. While a public API that exposes an immutable set is preferable, this is a compromise since no such set exists. Alternatives not used are: `scala.collection.Set`, `scala.collection.mutable.LinkedHashSet`, and `ListSet`.

A `DependencyManager` can then be used to, given a set of transforms you want to run and a set of all transforms that have been run, determine a linear ordering of transforms that satisfy your requirements. This uses a modification of the algorithm proposed by @azidar in https://github.com/freechipsproject/firrtl/issues/446#issuecomment-300573107:

1. A DAG of transforms with *invalidation edges* is constructed (an *invalidates graph*)
2. A DAG of transforms with *prerequisite and dependent edges* is constructed (a *dependents graph*)
3. A toplogical sort of the dependents graph, seeded with the reverse topological sort of the invalidates graph, gives a good ordering of transforms that minimizes re-lowerings
4. This ordering is examined, node by node, cleaning up any mismatches between transforms by solving `DependencyManager` sub-problems

This PR adds the `seededLinearize` method to `DiGraph` that allows for a topological sort seeded with some starting nodes. This is required to implement the above algorithm.

Finally, in preparation for using `DependencyManager` for FIRRTL transform scheduling, this makes `Transform` extend `TransformLike[CircuitState]`. 

META:
- Depends on #1005 
- Fixes #948 
- Related to #446 

## Examples

The examples below show examples of how different combinations of dependencies interact and are resolved. This uses built-in `DependencyManager` methods for generating graphviz output. All examples below are taken from the included `PhaseManagerSpec`. 

The top graph shows dependencies with prerequisites as black arrows, dependents as red arrows, and invalidates with dashed lines. The bottom graph shows the resulting linear ordering (if one exists). Each `DependencyManager` sub-problem is shown as a cluster with it's "target transforms" and the "state" showing current transforms that have run.

### Simple Dependency

Here, `B` depends on `A` and we request to run `B`.

dependencyGraph dot
transformOrder dot

### Simple Invalidate

Here, you have a pretty basic graph with one invalidate: `D` invalidates `A`, but also requires `A`. To get everything to run, you have to solve a sub-problem that reruns `A`.

dependencyGraph dot
transformOrder dot

### Single Dependent (Prerequisite Injection)

This is intended to model the @stevenmburns case of needing to run a custom transform *before* something else. Here, `Second` depends on `First`, but we want to run the custom transform `Custom` which should be inserted between them. By making `Second` a dependent of `Custom` you get the ordering you want.

dependencyGraph dot
transformOrder dot

### Two Invalidates

In this graph, there are only two invalidations. However, this teases out problems with sub-optimal linearizations (corresponding to different possible topological sorts of the dependents graph). 

dependencyGraph dot
transformOrder dot

Here is the resulting linearization without seeding using the invalidation graph.

transformOrder dot

### Complicated Example

The graph sum of the dependents and invalidates graphs form a cycle. This is a test with multiple valid orderings where choosing a bad one will result in a lot of re-lowering.

dependencyGraph dot
transformOrder dot

Similar to the above "Two Invalidates", you wind up with a valid, but worse solution without the seeded topological sort.

transformOrder dot

### Chained Invalidation

In this graph, you start with a simple initial graph with some weird properties where your invalidations start invalidating other needed transforms. This graph has these odd properties:

- `A` invalidates `B`
- `B` invalidates `C`
- `C` invalidates `D`
- `E` depends on `[A, B, C, D]`
- You start in state `[B, C, D]` wanting to run `[A, E]`

This tests that "chained invalidation" is getting properly handled.

dependencyGraph dot
transformOrder dot

### Cyclic Prerequisite

A cycle in the prerequisistes will throw an exception.

dependencyGraph dot

### Cyclic Invalidates

A cycle in the invalidates will throw an exception.

dependencyGraph dot

### Repeated Analyses

This covers an intended use case where some "analysis" phase is repeatedly invalidated by some "consumer" phases. Concretely an analysis phase would be a phase that produces information that a consumer phase wants to use. However, each consumer phase makes modifications which require the analysis phase to be re-run (e.g., an analysis phase that constructs an `InstanceGraph` and consumer phases that rename modules).

dependencyGraph dot
transformOrder dot

### Inverted Repeated Analyses

This is equivalent to the repeated analyses above, but the consumer phases invalidate each other.

dependencyGraph dot
transformOrder dot

### Deterministic Order

This relatively big example asserts that a deterministic order is preserved for all graphs constructed inside a `DependencyManager`.

dependencyGraph dot
transformOrder dot

### CODE REVIEW

**Comment by azidar** [comment]
It would be good to get @jackkoenig approval to ensure his workload works.

**Comment by seldridge** [comment]
I'm hitting issues with determinism due to use of `Set`. @jackkoenig suggestion is to use `Seq` and then convert internally.

**Comment by seldridge** [comment]
Several updates in an omnibus commit that should likely be broken up:

- A new `firrtl.stage.Forms` object is added that defines the old lowering compiler forms in terms of sets of transforms/passes that should be run. A cool thing to fall out of this is that higher forms are subsets of lower forms. You can then define an ordering that mirrors the old `CircuitForm.value` concept.
- The explicit transform/pass order in each lowering compiler is removed and replaced with a transform manager that solves a transform ordering sub-problem between two `firrtl.stage.Forms`.
- Dependencies are defined for all passes/transforms (used by the lowering compilers)
- A `LoweringCompilersSpec` is added that checks that the new transform ordering mostly matches the old. The "mostly" comes from additional checks being added.
- This fixes a few bugs in the way a dependency manager ordered things.

One interesting thing to fall out of this. There are a lot of transforms that you could deem optional: `CheckTypes`, `ConstantPropagation`, etc. If you define the dependencies of optional transforms using `dependents` as opposed to `prerequisites`, you can construct forms where they are used or not used. Consider the situation of `MinimumLowFirrtlOptimization` vs. `LowFirrtlOptimization`. Roughly, low firrtl optimization is the minimum optimizations with constant propagation and some other transforms mixed in. If you define the correct invalidates for transforms in min low and the correct dependents in constant propagation, you can just "mix in" constant propagation and everything works. I need to redefine all the checks this way, but this seems incredibly useful for composing superset forms from subset forms.

Also, the use of sets for defining forms gives you a nice ordering (using `_.subsetOf(_)`). This effectively mirrors the old `CircuitForm.value`. 

Stability problems of `Set` still not fixed...

**Comment by seldridge** [comment]
I updated this to clear a bug that Chisel tests exposed and move to a `Seq` for dependency specification. @jackkoenig had suggested using a `ListSet` for the latter instead. I started to migrate to this, but I think the reverse insertion order of `ListSet` may make things somewhat confusing for a user (which way are they expecting things to run?). As a middle ground, I opted to include a note in the scaladoc on methods that take a `Seq` to indicate the reason for this and that it'll be converted to an  ordered `Set` internally.

@jackkoenig: once things calm down for you, can you see if this works on SiFive workloads?

**Comment by seldridge** [comment]
I removed all code related to moving FIRRTL to use the Dependency API. This is a lot of additional code that makes the content of this PR get lost. I'm moving the port into a separate PR.

This PR is then a pure API addition that adds the `DependencyAPI` trait, defines `DependencyManager` and `PhaseManager`, and adds tests.

**Comment by sequencer** [comment]
I have a question that, will `Phase` finally deprecate `Transform`?
Since all `Transform` is modifying the
```
CircuitState(
    circuit: Circuit,
    form: CircuitForm,
    annotations: AnnotationSeq,
    renames: Option[RenameMap])
```
which is a wrapper of `AnnotationSeq` for `Compiler`.
Since `Phase` is a `TransformLike[AnnotationSeq]`, in order to decrease redundancy of `Phase`, will `Compiler` and `Transform` deprecate in the future, so we can easily manage dependency between `Phase` and `Transform`.

**Comment by seldridge** [comment]
@sequencer: That's a good question... At this point, I have not considered deprecating `Transform` in favor of `Phase`, nor has this come up in dev meetings. It could, however, be something worth discussing long term (FIRRTL 1.4). Part of the motivation of something like that is, as you mention, being able to define one `DependencyManager[A]` and have that deal with all dependencies as opposed to having to interface different `TransformLike`s together.

To elucidate the current approach ([FIRRTL#1105](https://github.com/freechipsproject/firrtl/pull/1105)) which uses a `DependencyManager[A]` (`firrtl.stage.FirrtlStage`) which internally calls a `Translator[A, B]` (`firrtl.stage.phases.Compiler`) which internally calls a `DependencyManager[B]` (`firrtl.stage.transforms.Compiler`). You can imply dependencies in this way, but it's not explicit. Maybe it's worth thinking about how we could do this...

Nevertheless, a full deprecation of `Transform` would cause a **ton** of pain... some of which I'm aware of:

- We'd need to have complete backwards compatibility for all existing third-party transforms. This likely wouldn't be too bad as you can always provide backwards compatibility for `Transform` in a `Translator[CircuitState, AnnotationSeq]`.
- The `AnnotationSeq` abstraction may not always be the best container, especially if there's something that better encapsulates the type over which you're doing a mathematical transformation. While I'm possibly motivated in ramming everything into this, it may not always be the best fit.

`AnnotationSeq` is really just a `Monad`-like pattern. While we can make this the de-facto standard, there's likely some benefit in letting users use the abstraction that makes sense for their compiler internals. E.g., abstractly, a FIRRTL transform is most natively expressed as `(Circuit) -> Circuit` and not `(AnnotationSeq) -> AnnotationSeq` even though it could be expressed that way. It's hard to know where to draw the line...

In summary, good thoughts, I'm super interested in this type of discussion, and am open to ideas. Stage/Phase and Dependency API are really about *trying to build **hardware compiler infrastructure*** and it's worth putting a lot of thought into this.

**Comment by sequencer** [comment]
Thank you for your replying. Really appreciate your answer!
In current status, it's seems easy to manage the dependency among different `Transform`, that help me a lot!
BTW, will you accept a `CustomPhaseAnotation` PR for `firrtl`? After `Dependency API` finally merged, I think it's easy to add custom `Phase` to do some additional `os` works. For example, emiting `hammer` config or directly calling the EDA like hammer does.

**Comment by seldridge** [comment]
Something like `CustomPhaseAnnotation` likely makes sense as does the motivation you state. It would need to have an API that includes specification of in which `Stage` you want it to run since, unlike with `Transform`s which are only associated with FIRRTL, a `Phase` could run in a lot of places (Chisel, FIRRTL, testers, etc.). I think this also means that a `Stage` needs to be a `DependencyManager` or be a very thin wrapper around one. Likely worth a separate issue or maybe on RFC.

Something like:
```scala
RunPhaseAnnotation(stage: Class[Stage], phase: Class[Phase])`
```

**Comment by seldridge** [comment]
After some confirmation from @jackkoenig that this passes SiFive workloads, this is now rebased and merged.
